Initial attempt at isolating resources between dependencies

stevendanna opened the pull request:
Previously, all resources were loaded into a single resource registry.
Now, each profile context has a resource registry, when a profile's
library is loaded into the profile context, we update the
profile-context-specific resource registry.  This local registry is
then used to populate the execution context that the rules are
evaluated in.

Signed-off-by: Steven Danna steve@chef.io

stevendanna committed:
Initial attempt at isolating resources between dependencies

Previously, all resources were loaded into a single resource registry.
Now, each profile context has a resource registry, when a profile's
library is loaded into the profile context, we update the
profile-context-specific resource registry.  This local registry is
then used to populate the execution context that the rules are
evaluated in.

Signed-off-by: Steven Danna <steve@chef.io>

chris-rock commented on test/unit/mock/profiles/dependencies/profile_c/libraries/gordon_config.rb:
```diff
@@ -0,0 +1,54 @@
+  # Load the configuration file on initialization
+  def initialize
+    @params = {}
+    @path = '/tmp/gordon/config.yaml'
```
we may simplify the resource to just return a version number. Otherwise we are relying on external configuration. eg.

```
def version
   1
end
```

chris-rock commented on test/unit/mock/profiles/dependencies/profile_c/libraries/gordon_config.rb:
```diff
@@ -0,0 +1,54 @@
+class GordonConfig < Inspec.resource(1)
+  name 'gordon_config'
+
+  desc "
```
we do not need desc and example in this test

chris-rock commented on test/unit/mock/profiles/dependencies/profile_a/controls/example.rb:
```diff
@@ -20,3 +20,9 @@
+
+control 'profilea-2' do
+  describe gordon_config do
+    its('version') { should eq('1.0') }
```
lets test for `its('version') { should eq 1 }`

stevendanna replied:
Simplified these as suggested.

chris-rock replied:
This looks great now

chris-rock commented on test/unit/mock/profiles/dependencies/profile_b/controls/example.rb:
```diff
@@ -14,3 +14,9 @@
+
+control 'profileb-2' do
+  describe gordon_config do
+    its('version') { should eq('1.0') }
```
lets test for `its('version') { should eq 2 }`

chris-rock commented on test/unit/mock/profiles/dependencies/inheritance/inspec.yml:
```diff
@@ -10,5 +10,3 @@ depends:
     path: ../profile_a
   - name: profile_b
     path: ../profile_b
-  - name: ssh-hardening
```
any reason why we are removing this?

stevendanna replied:
@chris-rock Mostly because the fetching wasn't the main thing we were working so the ssh output was really just cluttering up what we wanted to see.

chris-rock commented on lib/inspec/profile_context.rb:
```diff
@@ -4,17 +4,20 @@
 
-    attr_reader :attributes, :rules, :profile_id
+    attr_reader :attributes, :rules, :profile_id, :resource_registry
+    attr_reader :profile_execution_context
```
may it make sense to just use `execution_context`? since we are in the profile context already

stevendanna replied:
Based on our discussion I think we'll end up with two context, perhaps:
- `control_eval_context`
- `resource_eval_context`

chris-rock replied:
This makes sense to me

chris-rock commented on lib/inspec/resource.rb:
```diff
@@ -3,17 +3,36 @@
 
 module Inspec
   class Resource
+    REGISTRY_LOCK = Mutex.new
```
Why do we need threading and a mutex?

stevendanna replied:
Summary of discussion:

We can probably replace this by introducing a profile-specific class that all library files are evaluated against.

stevendanna committed:
Load all dependent libraries, even if include_context isn't called

The goal of these changes is to ensure that the libraries from
dependencies are loaded even if their controls are never included.  To
facilitate this, we break up the loading into seperate steps, and move
the loading code into the Profile which has acceess to the dependency
information.

Signed-off-by: Steven Danna <steve@chef.io>

chris-rock commented on test/unit/dsl/control_test.rb:
```diff
@@ -10,7 +10,7 @@ def load(content)
       'controls/mock.rb' => "control '1' do\n#{content}\nend\n",
     }
-    opts = { test_collector: Inspec::RunnerMock.new }
+    opts = { test_collector: Inspec::RunnerMock.new, backend: Inspec::Backend.create({}) }
```
wouldn't that require `Inspec::Backend.create({ backend: 'mock' })`?

chris-rock commented on lib/inspec/runner.rb:
```diff
@@ -102,61 +120,15 @@ def load_attributes(options)
-    # - `:controls`: An array of controls to run from this
-    #    profile. Any returned rules that are not part of these
-    #    controls will be filtered before being passed to @test_collector.
+    # @params _opts [Hash] Unused, but still here to avoid breaking kitchen-inspec
```
We should add a ticket in kitchen-inspec as a follow-up to fix that

stevendanna committed:
Use per-profile execution contexts for library loading

Previously, libraries were loaded by instance_eval'ing them against
the same execution context used for control files.  All resources were
registered against a single global registry when the `name` dsl method
was invoked.  To obtain seperation of resources, we would mutate the
instance variable holding the globale registry and then change it back
at the end.

Now, we instance_eval library files inside an anonymous class.  This
class has its own version of `Inspec.resource` that returns another
class with the resource DSL method and the profile-specific resource
registry.

chris-rock commented on lib/inspec/base_cli.rb:
```diff
@@ -70,7 +70,7 @@ def run_tests(targets, opts)
 
       runner = Inspec::Runner.new(o)
-      targets.each { |target| runner.add_target(target, opts) }
+      targets.each { |target| runner.add_target(target) }
```
👍

chris-rock commented on lib/inspec/control_eval_context.rb:
```diff
@@ -0,0 +1,139 @@
+        end
+
+        def to_s
+          "Profile Context Run #{profile_name}"
```
Is this really representing a full profile context? Its only representing one section aka file in a profile

chris-rock commented on lib/inspec/control_eval_context.rb:
```diff
@@ -0,0 +1,139 @@
+          register_control(rule_class.new(id, profile_id, opts, &block))
+        end
+
+        define_method :describe do |*args, &block|
```
do you think we should add a comment that this is simulating the rspec describe and implicitly generate a control?

chris-rock commented on lib/inspec/control_eval_context.rb:
```diff
@@ -0,0 +1,139 @@
+          res
+        end
+
+        define_method :add_resources do |context|
```
the methods `add_resources`, `add_subcontext` and `register_control` should not be available to our DSL right?

stevendanna replied:
We could hide these from the API directly via a layer of indirection.  If we want them to be completely hidden, I think we would need the Inspec::DSL code to change a bit more substantially.  Happy to do either.

chris-rock commented on lib/inspec/dependencies/dependency_set.rb:
```diff
@@ -18,14 +18,22 @@ class DependencySet
     # @param vendor_path [String] Path to the vendor directory
     #
-    def self.from_lockfile(lockfile, cwd, vendor_path)
+    def self.from_lockfile(lockfile, cwd, vendor_path, backend)
```
param is not defined in comments

chris-rock commented on lib/inspec/dependencies/dependency_set.rb:
```diff
@@ -18,14 +18,22 @@ class DependencySet
       vendor_index = VendorIndex.new(vendor_path)
       dep_tree = lockfile.deps.map do |dep|
-        Inspec::Requirement.from_lock_entry(dep, cwd, vendor_index)
+        Inspec::Requirement.from_lock_entry(dep, cwd, vendor_index, backend)
```
Somehow it feels strange that the dependency resolution needs a backend? Why is that required?

stevendanna replied:
Yeah, I agree this is gross.  The reason we started threading this through everything is that we need the backend anywhere we might instantiate a ProfileContext.  Since we instantiate ProfileContext's as we are building Inpec::Profile's, we now need it all over the place.

chris-rock replied:
As discussed, lets improve that in our next iterations

chris-rock commented on lib/inspec/dsl_shared.rb:
```diff
@@ -0,0 +1,25 @@
+  # Contains methods we would like in multiple DSL
+  #
+  module DSL
+    module RequireOverride
```
nice separation

stevendanna committed:
Minor code-review change to ControlEvalContext

- Add comment describing describe
- Fixup to_s output to be more accurate

Signed-off-by: Steven Danna <steve@chef.io>
